File: backend/src/main/java/dev/frilly/locket/data/Comment.java
package dev.frilly.locket.data;

import jakarta.persistence.*;

import java.time.LocalDateTime;

/**
 * Entity model for a comment under a post.
 */
@Entity
@Table(name = "comments")
public final class Comment {

  @Column(nullable = false)
  private final LocalDateTime time;

  @ManyToOne
  @JoinColumn(name = "user_id", nullable = false)
  private User user;

  @ManyToOne
  @JoinColumn(name = "post_id", nullable = false)
  private Post post;

  @Column(nullable = false, columnDefinition = "text")
  private String content;
  
  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private long id;

  public Comment(User user, Post post, String content) {
    this();
    this.user    = user;
    this.post    = post;
    this.content = content;
  }

  public Comment() {
    time = LocalDateTime.now();
  }

  public long id() {
    return id;
  }

  public User user() {
    return user;
  }

  public Post post() {
    return post;
  }

  public String content() {
    return content;
  }

  public LocalDateTime time() {
    return time;
  }

}

File: backend/src/main/kotlin/dev/frilly/locket/data/Comment.kt
package dev.frilly.locket.data

import jakarta.persistence.*
import java.time.LocalDateTime

/**
 * Represents a comment under a post. Should this be separated from messages?
 *
 * The table has:
 * - ID (Serial) primary key
 * - USER_ID (Serial) references users
 * - POST_ID (Serial) references posts
 * - CONTENT (Text) the comment
 * - TIME (Time) when the comment was posted
 */
@Entity
@Table(name = "comments")
data class Comment(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long,

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @ManyToOne
    @JoinColumn(name = "post_id", nullable = false)
    val post: Post,

    @Column(nullable = false, columnDefinition = "text")
    val content: String,

    @Column(nullable = false)
    val time: LocalDateTime = LocalDateTime.now(),
)
